### PR TITLE
snap: add io.snapcraft.Settings to `snap userd`

### PR DESCRIPTION
**Pull request by mvo5**
This PR implements the idea of having support for xdg-settings inside snaps. It works similar to xdg-open. I.e. we create a xdg-settings inside the core snap (just like we do now with xdg-open) and send the requests via dbus to the host. Threre `snapd userd` can gather the data, ask the user and send the result back. This should allow e.g. firefox/chrome to set the default browser (as requested e.g.  here: https://forum.snapcraft.io/t/browser-not-setting-itself-as-the-default/1547/14).

If this looks like the right path I will add the required spread tests around it etc. There is also the open question if we want to use tools like zenity/kdialog for the user questions or if we want to use one of the native bindings to gtk/qt. And if the later, which one :)

The core snap side if here: https://github.com/snapcore/core/pull/62

### CODE REVIEW

**Comment by zyga** [comment]
Thank you for making this proposal @mvo5. I had a look at the code and I was expecting to find a pass-through that would grab a request and send it to whatever implements the real `xdg-settings` on the desktop. Unless the original API is safe the point of the passthrough is to implement a filter that rejects certain requests. Do you envision having a filter here?

**Comment by mvo5** [comment]
@zyga Yes, I think we want to get input from @jdstrand how this should restrict usage. Setting is currently always ask the user. However we need to discuss what to do about check/get. Either a whitelist or a question or maybe even unrestricted if the information leak is harmless.

File: userd/launcher.go [hunk 1]
```diff
@@ -54,6 +54,11 @@ func (s *Launcher) Name() string {
 	return "io.snapcraft.Launcher"
 }
 
+// BasePath returns the base path of the object
+func (s *Launcher) BasePath() dbus.ObjectPath {
+	return "/io/snapcraft/Launcher"
+}
```
Comment (jdstrand) [inline userd/launcher.go]: This seems unrelated to this PR?

File: userd/settings.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+	return "/io/snapcraft/Settings"
+}
+
+// IntrospectionData gives the XML formatted introspection description
+// of the DBus service.
+func (s *Settings) IntrospectionData() string {
+	return settingsIntrospectionXML
+}
+
+// Check implements the 'Check' method of the 'com.canonical.Settings'
```
Comment (jdstrand) [inline userd/settings.go]: s/com.canonical.Settings/io.snapcraft.Settings/

File: userd/settings.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+//
+// Example usage: dbus-send --session --dest=io.snapcraft.Settings --type=method_call --print-reply /io/snapcraft/Settings io.snapcraft.Settings.Check string:'default-web-browser' string:'firefox.desktop'
+func (s *Settings) Check(setting, check string) (string, *dbus.Error) {
+	cmd := exec.Command("xdg-settings", "check", setting, check)
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return "", dbus.MakeFailedError(fmt.Errorf("cannot check setting %s: %s", setting, osutil.OutputErr(output, err)))
+	}
+
+	return strings.TrimSpace(string(output)), nil
```
Comment (jdstrand) [inline userd/settings.go]: The whole check/get/set paradigm doesn't map directly to snaps wrt info-disclosure. I think we can handle this though since userd can mediate.
* if 'Check()', then see if the value of 'check' matches one of the desktop files of the snap. If not, return error, if so, send it along to the system's xdg-settings and return the result
* likewise if Set(), verify if 'new' belongs to the snap. If not, return an error, if so, send it along to the system's xdg-settings and return the result
* Get() is slightly harder but doable. Run the get request on the system's xdg-settings and see if the result belongs to the snap, if so, return the result, if not, mock up something (eg, `NOT_THIS_SNAP.snap.desktop` (or similar-- I chose uppercase and an extra '_' because that is an invalid snap name/command combination for a desktop file)).

File: userd/settings.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+func (s *Settings) IntrospectionData() string {
+	return settingsIntrospectionXML
+}
+
+// Check implements the 'Check' method of the 'com.canonical.Settings'
+// DBus interface.
+//
+// Example usage: dbus-send --session --dest=io.snapcraft.Settings --type=method_call --print-reply /io/snapcraft/Settings io.snapcraft.Settings.Check string:'default-web-browser' string:'firefox.desktop'
+func (s *Settings) Check(setting, check string) (string, *dbus.Error) {
+	cmd := exec.Command("xdg-settings", "check", setting, check)
```
Comment (jdstrand) [inline userd/settings.go]: We should check 'setting' here and in Get() and Set() against known keys we want to support. In this case, `default-web-browser`. xdg-settings also supports `default-url-scheme-handler`, but I don't think that is used much (it isn't set here). Perhaps wait until someone asks for it?

Furthermore, since the snap will necessarily have to be able to use these methods, we need to fastidiously verify all input coming in on dbus.

File: userd/settings.go (hunk 3 above)
Comment (jdstrand) [reply userd/settings.go]: The question then becomes how to answer what snap the connecting process is coming from. You can either do this with libapparmor or see which snap cgroup freezer the pid is in. The former is easier to implement and most correct, but the latter works cross-distro where apparmor isn't available.

**Comment by jdstrand** [comment]
I might also mention that there is xdg-mime, which has also been requested: https://forum.snapcraft.io/t/unable-to-set-default-mail-client-x-scheme-handler-in-snap-desktop-files-are-ignored/2543

File: userd/launcher.go (hunk 1 above)
Comment (mvo5) [reply userd/launcher.go]: This is prereq work (part of 585449b4f06b5bffea5c14fed3bd8951f7520e4f) to generalize the dbusInterface to support multiple interfaces at the same time. I can propose this as a separate PR if desired.

File: userd/settings.go (hunk 2 above)
Comment (mvo5) [reply userd/settings.go]: Thanks, indeed, this is wrong in a number of places. All fixed now.

File: userd/settings.go (hunk 4 above)
Comment (mvo5) [reply userd/settings.go]: Sounds good, I added a whitelist to only allow default-web-browser for now.

File: userd/settings.go (hunk 3 above)
Comment (mvo5) [reply userd/settings.go]: Silly question - but how terrible is the information disclosure we have here in Get/Check? Given that we only allow to query for "default-web-browser" currently and given that the answer is a string like "firefox.desktop" I wonder if that is an acceptable risk? As for Set() this is currently behind a dialog so no snap can sneak behind the users back. But I see that for Set() we may want to protect things a bit more. I can see a (weak) use case for a browser where e.g. the browser has "set $browser as default" and if the user sets it a option "revert to previous setting" where it does make sense to allow a snap to set something else than itself.

**Comment by mvo5** [comment]
I pushed a integration test now as well, however that will only work once https://github.com/snapcore/core/pull/62 is merged and in the edge channel.

File: userd/launcher.go (hunk 1 above)
Comment (jdstrand) [reply userd/launcher.go]: No need to do that. Thanks for the explanation. :)

File: userd/settings.go (hunk 3 above)
Comment (jdstrand) [reply userd/settings.go]: The information leak is not terrible in this case, but avoidable. One day I'd like to be in a position where there or no information leaks, so I try to avoid introducing any new whenever feasible.

**Comment by zyga** [comment]
@mvo5 snapcore/core#62 is now merged.

File: userd/launcher_test.go [hunk 5]
```diff
@@ -28,6 +30,8 @@ import (
 	"github.com/snapcore/snapd/userd"
 )
 
+func Test(t *testing.T) { TestingT(t) }
```
Comment (zyga) [inline userd/launcher_test.go]: Ouch :)

File: userd/settings.go [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+</interface>
+<interface name='io.snapcraft.Settings'>
+	<method name='Check'>
+		<arg type='s' name='setting' direction='in'/>
+		<arg type='s' name='check' direction='in'/>
+                <arg type='s' name='result' direction='out'/>
+	</method>
+	<method name='Get'>
+		<arg type='s' name='setting' direction='in'/>
+                <arg type='s' name='result' direction='out'/>
```
Comment (zyga) [inline userd/settings.go]: Space off by one here and above

File: userd/settings.go [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+	</method>
+	<method name='GetMachineId'>
+               <arg type='s' name='machine_uuid' direction='out'/>
+	</method>
+</interface>
+<interface name='io.snapcraft.Settings'>
+	<method name='Check'>
+		<arg type='s' name='setting' direction='in'/>
+		<arg type='s' name='check' direction='in'/>
+                <arg type='s' name='result' direction='out'/>
```
Comment (zyga) [inline userd/settings.go]: Space off by one here and below.

Comment (mvo5) [reply userd/settings.go]: tab vs space, fixing.

**Comment by mvo5** [comment]
Note this needs https://github.com/snapcore/core/pull/65

File: userd/helpers.go [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+	}
+	call.Store(&pid)
+	return pid, nil
+}
+
+// FIXME: move to osutil?
+func snapFromPid(pid int) (string, error) {
+	// racy :( - maybe move to use libapparmor but that won't work on
+	// every distro yet
+	m, err := filepath.Glob(filepath.Join(dirs.FreezerCgroupDir, "snap.*"))
```
Comment (bboozzoo) [inline userd/helpers.go]: This should be avaialble in proc under `/proc/<pid>/cgroup`. For example for `snap run --shell hello-world`:
```
$ cat /proc/5873/cgroup 
10:devices:/user.slice
9:cpuset:/
8:net_cls,net_prio:/
7:freezer:/snap.hello-world
6:perf_event:/
5:pids:/user.slice/user-1000.slice/user@1000.service
4:cpu,cpuacct:/
3:memory:/
2:blkio:/
1:name=systemd:/user.slice/user-1000.slice/user@1000.service/gnome-terminal-server.service
0::/user.slice/user-1000.slice/user@1000.service/gnome-terminal-server.service
```

File: userd/settings.go [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+
+// Check implements the 'Check' method of the 'io.snapcraft.Settings'
+// DBus interface.
+//
+// Example usage: dbus-send --session --dest=io.snapcraft.Settings --type=method_call --print-reply /io/snapcraft/Settings io.snapcraft.Settings.Check string:'default-web-browser' string:'firefox.desktop'
+func (s *Settings) Check(setting, check string, sender dbus.Sender) (string, *dbus.Error) {
+	// avoid information leak: see https://github.com/snapcore/snapd/pull/4073#discussion_r146682758
+	snap, err := snapFromSender(s.conn, sender)
+	if err != nil {
+		dbus.MakeFailedError(err)
```
Comment (bboozzoo) [inline userd/settings.go]: Use the error?

File: userd/settings.go [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+	cmd := exec.Command("xdg-settings", "get", setting)
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return "", dbus.MakeFailedError(fmt.Errorf("cannot get setting %s: %s", setting, osutil.OutputErr(output, err)))
+	}
+
+	// avoid information leak: see https://github.com/snapcore/snapd/pull/4073#discussion_r146682758
+	snap, err := snapFromSender(s.conn, sender)
+	if err != nil {
+		dbus.MakeFailedError(err)
```
Comment (bboozzoo) [inline userd/settings.go]: the error is not used?

File: userd/settings.go [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+	}
+
+	cmd := exec.Command("xdg-settings", "get", setting)
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return "", dbus.MakeFailedError(fmt.Errorf("cannot get setting %s: %s", setting, osutil.OutputErr(output, err)))
+	}
+
+	// avoid information leak: see https://github.com/snapcore/snapd/pull/4073#discussion_r146682758
+	snap, err := snapFromSender(s.conn, sender)
```
Comment (bboozzoo) [inline userd/settings.go]: I think this check should be done before `exec.Command(...)`.

File: userd/settings.go [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+		return err
+	}
+	if !allowedSettingsValues.MatchString(new) {
+		return dbus.MakeFailedError(fmt.Errorf("cannot set setting %q to value %q: value not allowed", setting, new))
+	}
+
+	// see https://github.com/snapcore/snapd/pull/4073#discussion_r146682758
+	snap, err := snapFromSender(s.conn, sender)
+	if err != nil {
+		dbus.MakeFailedError(err)
```
Comment (bboozzoo) [inline userd/settings.go]: Returned error is not  used.

File: userd/settings.go (hunk 9 above)
Comment (mvo5) [reply userd/settings.go]: *cough* thank you - I'm a muppet

File: userd/settings.go (hunk 11 above)
Comment (mvo5) [reply userd/settings.go]: This check depends on the output from the exec.Command() call. Or am I missing something?

File: userd/helpers.go (hunk 8 above)
Comment (mvo5) [reply userd/helpers.go]: Thanks, I have a look if it this is easier to parse than the current (indirect) appraoch.

File: userd/settings.go [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+	</method>
+</interface>`
+
+// FIXME: allow setting default-url-scheme-handler ?
+var settingsWhitelist = []string{
+	"default-web-browser",
+}
+
+// FIXME: too restrictive?
+var allowedSettingsValues = regexp.MustCompile(`^[a-zA-Z0-9._]+$`)
```
Comment (jdstrand) [inline userd/settings.go]: This should allow `<snap command>.desktop` I think (and no `_`). validate.go has validAppName and https://github.com/snapcore/snapd/pull/4399 has some code to add a new ValidAppName() that you could use. Eg, check if ends with `.desktop`, the use the helpers after splitting on '_' (and '.').

File: userd/settings.go [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+	if !allowedSettingsValues.MatchString(new) {
+		return dbus.MakeFailedError(fmt.Errorf("cannot set setting %q to value %q: value not allowed", setting, new))
+	}
+
+	// see https://github.com/snapcore/snapd/pull/4073#discussion_r146682758
+	snap, err := snapFromSender(s.conn, sender)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+	new = fmt.Sprintf("%s_%s", snap, new)
```
Comment (jdstrand) [inline userd/settings.go]: I don't see where we are verifying that `new` is from the snap, which would be good for usability. Eg, if snap `foo` calls xdg-settings with `nonexistent.desktop`, then the snap can set the browser to `foo_nonexistent.desktop`.

 The fact that you are prepending the snap name is excellent, thank you.

File: userd/settings.go [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+		return "", dbus.MakeFailedError(fmt.Errorf("cannot get setting %s: %s", setting, osutil.OutputErr(output, err)))
+	}
+
+	// avoid information leak: see https://github.com/snapcore/snapd/pull/4073#discussion_r146682758
+	snap, err := snapFromSender(s.conn, sender)
+	if err != nil {
+		return "", dbus.MakeFailedError(err)
+	}
+	if !strings.HasPrefix(string(output), snap+"_") {
+		return "NOT_THIS_SNAP.snap.desktop", nil
```
Comment (jdstrand) [inline userd/settings.go]: I wonder if this string would be better: `NOT-THIS-SNAP.desktop`. In this manner, it follows the same pattern as an allowed result (ie `<command>.desktop`, see below) which might help consumers with parsing the result.

Thank you for implementing this check! :)

File: userd/settings.go [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+	new = fmt.Sprintf("%s_%s", snap, new)
+
+	// FIXME: what GUI toolkit to use?
+	// FIXME2: we could support kdialog here as well
+	if !osutil.ExecutableExists("zenity") {
+		return dbus.MakeFailedError(fmt.Errorf("cannot find zenity"))
+	}
```
Comment (jdstrand) [inline userd/settings.go]: I don't see dependency changes in the packaging; should we add it?

File: userd/helpers.go (hunk 8 above)
Comment (mvo5) [reply userd/helpers.go]: I did this now, its slightly shorter and for bonus points I added a test

File: userd/settings.go (hunk 16 above)
Comment (mvo5) [reply userd/settings.go]: I added a recommends, I think we want to support the kde, xfce etc equivalents as well

File: userd/settings.go (hunk 14 above)
Comment (mvo5) [reply userd/settings.go]: Thanks, fixed now.

File: userd/settings.go [hunk 17]
```diff
@@ -0,0 +1,240 @@
...
+		return "", dbus.MakeFailedError(fmt.Errorf("cannot get setting %s: %s", setting, osutil.OutputErr(output, err)))
+	}
+
+	// avoid information leak: see https://github.com/snapcore/snapd/pull/4073#discussion_r146682758
+	snap, err := snapFromSender(s.conn, sender)
+	if err != nil {
+		return "", dbus.MakeFailedError(err)
+	}
+	if !strings.HasPrefix(string(output), snap+"_") {
+		return "NOT-THIS-SNAP.snap.desktop", nil
```
Comment (jdstrand) [inline userd/settings.go]: Did you mean for this to be `NOT-THIS-SNAP.snap.desktop` or `NOT-THIS-SNAP.desktop`?

File: userd/helpers.go [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+
+// FIXME: move to osutil?
+func snapFromPid(pid int) (string, error) {
+	f, err := os.Open(fmt.Sprintf("%s/proc/%d/cgroup", dirs.GlobalRootDir, pid))
+	if err != nil {
+		return "", err
+	}
+	defer f.Close()
+
+	scanner := bufio.NewScanner(f)
```
[the remark this answers is not in the document]
Comment (mvo5) [reply userd/helpers.go]: Updated!

File: userd/settings.go (hunk 17 above)
Comment (mvo5) [reply userd/settings.go]: Indeed, silly me. Thanks, fixed!
